cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+If there are enough resources, manual file review should be conducted in a sandboxed environment before releasing the files to the public.
+
+Adding some automation to the review could be helpful, which is a harsh process and well studied before its usage. Some services (_e.g._ Virus Total) provide APIs to scan files against well known malicious file hashes. Some frameworks can check and validate the raw content type and validating it against predefined file types, such as in [ASP.NET Drawing Library](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.imaging.imageformat)
+
+### File Storage Location
+
+The location where the files should be stored must be chosen based on security and business requirements. The following points are set by security priority, and are inclusive:
+
+1. Store the files on a **different host**, which allow for complete segragation of duties between the application serving the user, and the host handling file uploads and their storage.
+2. Store the files **outside the webroot**, where only administrative access is allowed.
+3. Store the files **inside the webroot**, and set them in write permissions only.
+   - If read access is required, setting proper controls is a must (_e.g._ internal IP, authorized user, etc.)
+4. Store the files to a **database**, most preferrably as Blobs in a NoSQL database. The database should be chosen based on the application's requirements.
+
+> Some files are emailed on point, or processed and dismissed, without being stored. It is essential to conduct the security measures discussed in this sheet before doing any actions on them.
+
+### User Permissions
+
+Before any file upload service is accessed, proper validation should occur on two levels for the user uploading a file:
+
+- Authentication level
+  - The user should be a registered user, or an identifiable user, in order to set restrictions and limitations for their upload capabilities
+- Authorization level
+  - Make sure that the user's accesses allow them to upload a file, or multiple files for administrative cases.
+
+### System Permissions
+
+Files should be stored in a way that ensures:
+
+- Allowed system users are the only ones capable of reading the files
+- Required modes only are set for the file
+  - If execution is required, scanning the file before running it is required as a security best practice, to ensure that no macros or hidden scripts are available.
+
+### Upload and Download Limits
+
+The application should set proper size limits for the upload service in order to protect the file storage capacity. The file size limit should be considered after file decompression is conducted, and by using secure methods to calculate zip files size. For more on this, see how to [Safely extract files from ZipInputStream](https://wiki.sei.cmu.edu/confluence/display/java/IDS04-J.+Safely+extract+files+from+ZipInputStream), Java's input stream to handle ZIP files.
```

Comment:
Most of the examples given are for Microsoft/.NET, so it seems to to now just give a Java example. I'd either remove this, or replace it with a more link to a more general discussion of the issue.

Reply:
We can't cover everything, yet it doesn't hurt to actually have a certain language safe solution. Developers can understand what's happening and actually replicate that behavior. There are not so many public discussions on file safety